fv3gfs/util/buffer.py

```diff
@@ -1,17 +1,104 @@
-from typing import Callable, Iterable, Optional, Dict, Tuple
+from typing import Callable, Generator, Iterable, Optional, Dict, Tuple, List
 from ._timing import Timer, NullTimer
-from numpy import ndarray
+import numpy as np
+from numpy.lib.index_tricks import IndexExpression
 import contextlib
-
-from .utils import assign_array, is_c_contiguous
+from .utils import (
+    is_c_contiguous,
+    safe_assign_array,
+    device_synchronize,
+    mpi_safe_allocator,
+)
 from .types import Allocator
 
+BufferKey = Tuple[Callable, Iterable[int], type]
+BUFFER_CACHE: Dict[BufferKey, List["Buffer"]] = {}
```

Comment:
In a later PR, we talked about making this a class and adding a __del__ method to Buffer objects that raises an exception unless the garbage collection is happening within the `BufferCache.__del__`. This would add safety to ensure we always push buffers back to the cache.

Reply:
Ticketed DSL-545